src/raster/mdarray.rs

```diff
@@ -0,0 +1,331 @@
+            // }
+            let mut dimensions: Vec<Dimension> = Vec::new();
+            let mut i = 0;
+            while i < n_dimension {
+                let ptr = c_dimensions.add(i);
+                let next = ptr.read();
+                let value = Dimension::from_c_dimension(self, next);
+                i += 1;
+                dimensions.push(value);
+            }
+            Ok(dimensions)
+        }
+    }
+
+    pub fn read_into_slice<T: Copy + GdalType>(
+        &self,
+        buffer: &mut [T],
+        array_start_index: Vec<u64>,
+        count: Vec<usize>,
+    ) -> Result<()> {
+        // let array_start_index = [array_start_index.0, array_start_index.1];
+        // let count =  [count.0, count.1];
+        let array_step: *const i64 = std::ptr::null();
+        let buffer_stride: *const i64 = std::ptr::null();
+        let data_type = unsafe { GDALMDArrayGetDataType(self.c_mdarray) };
+        let p_dst_buffer_alloc_start: *mut libc::c_void = std::ptr::null_mut();
+        let n_dst_buffer_alloc_size = 0;
+
+        let rv = unsafe {
+            gdal_sys::GDALMDArrayRead(
+                self.c_mdarray,
+                array_start_index.as_ptr(),
+                count.as_ptr(),
+                array_step,
+                buffer_stride,
+                data_type,
+                buffer.as_mut_ptr() as *mut c_void, // pDstBuffer: *mut libc::c_void,
+                p_dst_buffer_alloc_start,           // pDstBufferAllocStart: *const libc::c_void,
+                n_dst_buffer_alloc_size,
+            )
+        };
+
+        if rv != 1 {
+            return Err(_last_cpl_err(rv.try_into().unwrap())); // this is probably incorrect!
+        }
+
+        Ok(())
+    }
+
+    /// Read a 'Buffer<T>' from this band. T implements 'GdalType'
+    ///
+    /// # Arguments
+    /// * array_start_index - Values representing the starting index to read in each dimension (in [0, aoDims[i].GetSize()-1] range). Array of GetDimensionCount() values. Must not be nullptr, unless for a zero-dimensional array.
+    /// * count - Values representing the number of values to extract in each dimension. Array of GetDimensionCount() values. Must not be nullptr, unless for a zero-dimensional array.
+    pub fn read_as<T: Copy + GdalType>(
+        &self,
+        array_start_index: Vec<u64>,
+        count: Vec<usize>,
+    ) -> Result<Vec<T>> {
+        let pixels: usize = count.iter().product();
+        let mut data: Vec<T> = Vec::with_capacity(pixels);
+
+        // Safety: the read_into_slice line below writes
+        // exactly pixel elements into the slice, before we
+        // read from this slice. This paradigm is suggested
+        // in the rust std docs
+        // (https://doc.rust-lang.org/std/vec/struct.Vec.html#examples-18)
+        unsafe {
+            data.set_len(pixels);
+        };
+        self.read_into_slice(&mut data, array_start_index, count)?;
```

Comment:
In the example, they read into the array-ptr and then set the length. Does that make any difference?